backend/cn/lib/builtins.ml

```diff
@@ -135,41 +167,66 @@ let array_to_list_def =
 
 
 let is_null_def =
-  ( "is_null",
-    Sym.fresh_named "is_null",
-    mk_arg1 (fun p loc' ->
-      IT.sterm_of_term IT.(eq_ (IT.term_of_sterm p, null_ loc') loc')) )
+  mk_builtin_arg1
+    "is_null"
+    CN_loc
+    (fun p1 ->
+      Some
+        (CNExpr
+           ( builtin_location,
+             CNExpr_binop
+               ( CN_equal,
+                 CNExpr (builtin_location, CNExpr_var p1),
+                 CNExpr (builtin_location, CNExpr_const CNConst_NULL) ) )))
+    CN_bool
 
 
 let ptr_eq_def =
-  ( "ptr_eq",
-    Sym.fresh_named "ptr_eq",
-    mk_arg2 (fun (p1, p2) loc' ->
-      IT.(sterm_of_term @@ eq_ (term_of_sterm p1, term_of_sterm p2) loc')) )
+  mk_builtin_arg2
+    "ptr_eq"
+    ~arg1type:CN_loc
+    ~arg2type:CN_loc
+    (fun p1 p2 ->
+      Some
+        (CNExpr
+           ( builtin_location,
+             CNExpr_binop
+               ( CN_equal,
+                 CNExpr (builtin_location, CNExpr_var p1),
+                 CNExpr (builtin_location, CNExpr_var p2) ) )))
+    CN_bool
 
 
 let prov_eq_def =
```

Comment:
This `prov_eq` function wasn't used anywhere in the tests. What is the difference between this, `ptr_eq` and `addr_eq`

Reply:
It's related to memory-model aspects of the CN (in case you're unfamiliar: a memory model is the part of the contract between programmers and compiler writers which sets out the rules for the latter to follow, and defines the envelope of optimisations for the former to engage in, relating to the rules around creating, manipulating and destroying pointers and memory). A memory model of C that seems to have some consensus is called PNVI-ae-udi, but the one CN will use (as soon as I'm done implementing it) is based on VIP.

If you can keep it for now, that would be helpful, because I may need it once I finish up other implementations and turn on VIP by default. You can put a comment next to it saying it should be deleted if it's not used after CN switches to enabling VIP by default.